ilp-core/src/main/java/org/interledger/InterledgerAddress.java

```diff
@@ -1,5 +1,7 @@
 package org.interledger;
 
+import org.immutables.value.Value;
```

Comment:
First of all, this all looks great - nice work! (this has actually been on my todo list for a while).

One thing I've noticed with Immutables is that if you annotate the interface, then you can't extend that interface with another Immutable. In order to do that, we need to use an `abstract` class. It's a _little_ more code for us, but I think we should do it so that somebody could create an extension of our interfaces, but still be able to use the Immutables library. 

One nice effect this has, however, is it allows us to define the `equals` and `hashcode` methods correctly, so that anybody using the library doesn't have to worry about it.

So, `InterledgerAddress.java` would instead look like this:

```
package org.interledger;

import org.immutables.value.Value;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Interledger Protocol (ILP) Addresses identify Ledger accounts (or groups of Ledger accounts) in
 * an ILP network, and provide a way to route a payment to its intended destination.
 *
 * <p>Interledger Addresses can be subdivided into two categories:</p>
 *
 * <p> <b>Destination Addresses</b> are complete addresses that can receive payments. A destination
 * address always maps to one account in a ledger, though it can also provide more specific
 * information, such as an invoice ID or a sub-account. Destination addresses MUST NOT end in a
 * period (.) character. </p>
 *
 * <p> <b>Address Prefixes</b> are incomplete addresses representing a grouping of destination
 * addresses. Many depths of grouping are possible, for example: groups of accounts or sub-accounts;
 * an individual ledger or sub-ledger; or entire neighborhoods of ledgers. Address prefixes MUST end
 * in a period (.) character. </p>
 *
 * <p> The formal specification for an Interledger Addresses is defined in <a
 * href="https://github.com/interledger/rfcs/tree/master/0015-ilp-addresses">Interledger RFC
 * #15</a>. </p>
 *
 * @see "https://github.com/interledger/rfcs/tree/master/0015-ilp-addresses"
 */
public interface InterledgerAddress {

  String REGEX = "(?=^.{1,1023}$)"
      + "^(g|private|example|peer|self|test[1-3])[.]([a-zA-Z0-9_~-]+[.])*([a-zA-Z0-9_~-]+)?$";

  Pattern PATTERN = Pattern.compile(REGEX);

  /**
   * <p>Constructor to allow quick construction from a String representation of an ILP address.</p>
   *
   * @param value String representation of an Interledger Address
   *
   * @return an {@link InterledgerAddress} instance.
   */
  static InterledgerAddress of(final String value) {
    return builder()
        .value(value)
        .build();
  }

  /**
   * <p>Helper method to determine if an Interledger Address conforms to the specifications outlined
   * in Interledger RFC #15.</p>
   *
   * @param value A {@link String} representing a potential Interledger Address value.
   *
   * @return {@code true} if the supplied {@code value} conforms to the requirements of RFC 15;
   *     {@code false} otherwise.
   *
   * @see "https://github.com/interledger/rfcs/tree/master/0015-ilp-addresses"
   */
  static boolean isValid(final String value) {
    Objects.requireNonNull(value);
    return PATTERN.matcher(value)
        .matches();
  }

  /**
   * Checks and requires that the specified {@code addressPrefix} is an address prefix per {@link
   * InterledgerAddress#isLedgerPrefix()}.
   *
   * <p>This method is designed primarily for doing parameter validation in methods and
   * constructors, as demonstrated below:</p> <blockquote>
   * <pre>
   * public Foo(InterledgerAddress bar) {
   *     this.ledgerPrefix = InterledgerAddress.requireAddressPrefix(bar);
   * }
   * </pre>
   * </blockquote>
   *
   * @param addressPrefix A {@link InterledgerAddress} to check.
   *
   * @return {@code ledgerPrefix} if its value ends with a dot (.).
   *
   * @throws IllegalArgumentException if the supplied Interledger address is not a ledger-prefix.
   */
  static InterledgerAddress requireAddressPrefix(final InterledgerAddress addressPrefix) {
    Objects.requireNonNull(addressPrefix, "addressPrefix must not be null!");
    if (!addressPrefix.isLedgerPrefix()) {
      throw new IllegalArgumentException(
          String.format("InterledgerAddress '%s' must be an Address Prefix ending with a dot (.)",
              addressPrefix.getValue()
          )
      );
    } else {
      return addressPrefix;
    }
  }

  /**
   * Checks and requires that the specified {@code addressPrefix} is not an address prefix per
   * {@link InterledgerAddress#isLedgerPrefix()}.
   *
   *
   * <p>This method is designed primarily for doing parameter validation in methods and
   * constructors, as demonstrated below:</p> <blockquote>
   * <pre>
   * public Foo(InterledgerAddress bar) {
   *     this.nonLedgerPrefix = InterledgerAddress.requireNotAddressPrefix(bar);
   * }
   * </pre>
   * </blockquote>
   *
   * @param addressPrefix A {@link InterledgerAddress} to check.
   *
   * @return {@code addressPrefix} if its value ends with a dot (.).
   *
   * @throws IllegalArgumentException if the supplied Interledger address is not a ledger-prefix.
   */
  static InterledgerAddress requireNotAddressPrefix(final InterledgerAddress addressPrefix) {
    Objects.requireNonNull(addressPrefix, "addressPrefix must not be null!");
    if (addressPrefix.isLedgerPrefix()) {
      throw new IllegalArgumentException(
          String
              .format("InterledgerAddress '%s' must NOT be an Address Prefix ending with a dot (.)",
                  addressPrefix.getValue()
              )
      );
    } else {
      return addressPrefix;
    }
  }

  /**
   * Get the default builder.
   *
   * @return a {@link ImmutableInterledgerAddress.Builder} instance.
   */
  static ImmutableInterledgerAddress.Builder builder() {
    return ImmutableInterledgerAddress.builder();
  }

  /**
   * Return this address's value as a non-null {@link String}. For example:
   * <code>us.usd.bank.account</code>
   *
   * @return A {@link String} representation of this Interledger address.
   */
  String getValue();

  /**
   * Tests if this Interledger address represents a ledger prefix.
   *
   * @return True if the address is a ledger prefix, false otherwise.
   */
  default boolean isLedgerPrefix() {
    return getValue().endsWith(".");
  }

  /**
   * Tests if this InterledgerAddress starts with the specified {@code addressSegment}.
   *
   * @param addressSegment An {@link String} prefix to compare against.
   *
   * @return {@code true} if this InterledgerAddress begins with the specified prefix.
   */
  default boolean startsWith(final String addressSegment) {
    Objects.requireNonNull(addressSegment, "addressSegment must not be null!");
    return this.getValue()
        .startsWith(addressSegment);
  }

  /**
   * Tests if this InterledgerAddress starts with the specified {@code interledgerAddress}.
   *
   * @param interledgerAddress An {@link InterledgerAddress} prefix to compare against.
   *
   * @return {@code true} if this InterledgerAddress begins with the specified prefix.
   */
  default boolean startsWith(final InterledgerAddress interledgerAddress) {
    Objects.requireNonNull(interledgerAddress, "interledgerAddress must not be null!");
    return this.startsWith(interledgerAddress.getValue());
  }

  /**
   * <p>Return a new InterledgerAddress by postfixing the supplied {@code segment} to this address.
   * </p>
   *
   * <p>This method can be used to construct both address prefixes and destination addresses. For
   * example, if the value of this address is '<code>us.usd.</code>', then calling this method with
   * an argument of '<code>bob</code>' would result in a new Interledger Address with a value of
   * '<code>us.usd.bob</code>', which is a destination address.</p>
   *
   * <p>Likewise, if the value of this address is '<code>us.usd.pacific.</code>', then calling this
   * method with an argument of '<code>creditunions.</code>' would result in a new Interledger
   * Address with a value of '<code>us.usd.pacific.creditunions.</code>', which is an address
   * prefix.</p>
   *
   * @param addressSegment A {@link String} to be appended to this address as an additional
   *                       segment.
   *
   * @return A new instance representing the original address with a newly specified final segment.
   */
  default InterledgerAddress with(String addressSegment) {
    Objects.requireNonNull(addressSegment, "addressSegment must not be null!");

    final StringBuilder sb = new StringBuilder(this.getValue());
    if (!this.isLedgerPrefix()) {
      sb.append(".");
    }
    sb.append(addressSegment);

    return InterledgerAddress.of(sb.toString());
  }

  /**
   * <p>Return this address's prefix, which is a new {@link InterledgerAddress} containing the
   * characters inside of {@link #getValue()}, up-to and including the last period. If this address
   * is already a prefix, then this instance is instead returned unchanged.</p>
   *
   * <p>For example, calling this method on an address 'g.example.alice' would yield a new address
   * containing 'g.example.'. Conversely, calling this method on an address that is already a
   * prefix, like 'g.example.' would yield the same instance, 'g.example.'.</p>
   *
   * @return A potentially new {@link InterledgerAddress} representing the prefix of this address.
   */
  default InterledgerAddress getPrefix() {
    if (this.isLedgerPrefix()) {
      return this;
    } else {
      return InterledgerAddress.of(getValue().substring(0, this.getValue().lastIndexOf(".") + 1));
    }
  }

  /**
   * <p>Return this address's parent prefix.</p>
   *
   * <p>If this address is a destination address, then this method returns a new {@link
   * InterledgerAddress} containing the characters inside of {@link #getValue()}, up-to and
   * including last period. If this address is instead a prefix, then this instance returns a new
   * {@link InterledgerAddress} containing the characters inside of {@link #getValue()}, up-to and
   * including second-to-last period, unless this address is a root prefix, in which case, {@link
   * Optional#empty()} is returned.</p>
   *
   * <p>For example, calling this method on an address 'g.example.alice' would yield a new address
   * containing 'g.example.'. Likewise, calling this method on an address that is already a prefix,
   * like 'g.example.' would yield 'g.'. Finally, calling this method on a root prefix, like "self."
   * would yield {@link Optional#empty()}.</p>
   *
   * @return An optionally present parent-prefix.
   */
  default Optional<InterledgerAddress> getParentPrefix() {
    // If this address is not a prefix, then just return the prefix. Otherwise, look deeper.
    if (this.isLedgerPrefix()) {
      // If the prefix is a root prefix, return Optional#empty. Otherwise, return the parent prefix.
      if (isRootPrefix()) {
        return Optional.empty();
      } else {
        // Call getParentPrefix with the account portion, which will forward to #getPrefix.
        return InterledgerAddress.of(
            this.getValue().substring(0, this.getValue().lastIndexOf("."))
        ).getParentPrefix();
      }
    } else {
      return Optional.of(this.getPrefix());
    }
  }

  /**
   * <p>Determines if this ILP Address has a parent-prefix.</p>
   *
   * <p>If this address is a destination address, then it has a parent prefix. However, if the
   * address is a prefix, then it only has a parent if it is _not_ a Root Prefix.</p>
   *
   * @return {@code true} if this address is a destination address. Otherwise (if this address is a
   *     prefix address), then return {@code false} if this address is a Root prefix; otherwise,
   *     return {@code true}.
   */
  default boolean hasParentPrefix() {
    // All ILP addresse have a parent prefix, except for Root prefixes.
    return isRootPrefix() == false;
  }

  String ROOT_REGEX = "^(g|private|example|peer|self|test[1-3])[.]$";
  Pattern ROOT_PATTERN = Pattern.compile(ROOT_REGEX);

  /**
   * <p>Determines if this address is a "root" prefix, which per ILP-RFC-15, is one of: "g.",
   * "private.", "example.", "peer.", "self.", "test1.", "test2.", or "test3.". Any other kind of
   * valid ILP address (e.g. "g.1") is not a root prefix.</p>
   *
   * @return {@code true} if this address is a root prefix; {@code false} otherwise.
   */
  default boolean isRootPrefix() {
    // Alternate implementation (Note: A prefix is a root prefix if it has only a single period)
    // final int numPeriods = getValue().length() - getValue().replaceAll("[.!?]+", "").length();
    // return numPeriods == 1;
    return ROOT_PATTERN.matcher(this.getValue()).matches();
  }

  @Value.Immutable
  abstract class AbstractInterledgerAddress implements InterledgerAddress {

    /**
     * <p>Compares the specified object with this <tt>InterledgerAddress</tt> for equality.</p>
     *
     * <p>The <tt>InterledgerAddress</tt> interface is essentially a type-safe wrapper around a
     * String value, so implementations should take care to forward equality decisions to the {@link
     * String#equals(Object)} method on the object returned by {@link #getValue()}.</p>
     *
     * @param obj object to be compared for equality with this collection
     *
     * @return <tt>true</tt> if the specified object is equal to this {@link InterledgerAddress}.
     *
     * @see Object#equals(Object)
     * @see Set#equals(Object)
     * @see List#equals(Object)
     */
    @Override
    public boolean equals(Object obj) {
      if (this == obj) {
        return true;
      }
      if (obj == null || getClass() != obj.getClass()) {
        return false;
      }

      InterledgerAddress foo = (InterledgerAddress) obj;

      return getValue().equals(foo.getValue());
    }

    /**
     * <p>Returns the hash code value for this <tt>InterledgerAddress</tt>.</p>
     *
     * <p>The<tt>InterledgerAddress</tt> interface is essentially a type-safe wrapper around a
     * String value, so implementations should take care to forward hashcode decisions to the {@link
     * String#equals(Object)} method on the object returned by {@link #getValue()}.</p>
     *
     * @return the hash code value for this {@link InterledgerAddress}.
     *
     * @see Object#hashCode()
     * @see Object#equals(Object)
     */
    @Override
    public int hashCode() {
      return this.getValue().hashCode();
    }

    /**
     * Precondition enforcer that ensures the value is a valid Interledger Address.
     *
     * @see "https://github.com/interledger/rfcs/blob/master/0015-ilp-addresses/0015-ilp-addresses.md"
     */
    @Value.Check
    void check() {
      if (!InterledgerAddress.isValid(getValue())) {
        throw new IllegalArgumentException(String.format("Invalid characters in address: ['%s']. "
            + "Reference Interledger ILP-RFC-15 for proper format.", getValue()));
      }
    }
  }
}
``` 

To make the above work, you'll need to edit `CodecContext#lookup` to be as follows:

```
/**
   * Helper method to lookup a {@link Codec} for the specified {@code type}.
   *
   * @param <T>  The specific type of {@link Codec} to return.
   * @param type An instance of {@link Class}.
   */
  @SuppressWarnings("unchecked")
  private <T> Codec<T> lookup(final Class<T> type) {
    Objects.requireNonNull(type);

    if (codecs.containsKey(type)) {
      return (Codec<T>) codecs.get(type);
    } else {
      // Check any interfaces...
      return Optional.ofNullable(
          Arrays.stream(type.getInterfaces())
              // Only use codecs that are found...
              .filter(codecs::containsKey)
              // If found, map to the first codec...
              .map(interfaceClass -> (Codec<T>) codecs.get(interfaceClass))
              .findFirst()
              // Otherwise, recurse with the super-class
              .orElseGet(() -> {
                // There were not interfaces, so check the super-classes.
                return (Codec<T>) lookup(type.getSuperclass());
              })
      ).orElseThrow(() -> new CodecException(
          String.format("No codec registered for %s or its super classes!",
              type.getName())));
    }
  }
```